File: model-client/src/main/java/org/modelix/model/api/ITree.java
package org.modelix.model.api;

import java.util.stream.LongStream;

public interface ITree {
  public static final long ROOT_ID = 1L;
  public static final String DETACHED_NODES_ROLE = "detached";

  void visitChanges(ITree oldVersion, ITreeChangeVisitor visitor);
  boolean containsNode(long nodeId);
  String getProperty(long nodeId, String role);
  LongStream getChildren(long parentId, String role);

  IConcept getConcept(long nodeId);

  long getParent(long nodeId);

  String getRole(long nodeId);

  ITree setProperty(long nodeId, String role, String value);

  INodeReference getReferenceTarget(long sourceId, String role);

  ITree setReferenceTarget(long sourceId, String role, INodeReference target);

  Iterable<String> getReferenceRoles(long sourceId);

  Iterable<String> getPropertyRoles(long sourceId);

  Iterable<String> getChildRoles(long sourceId);

  LongStream getAllChildren(long parentId);

  ITree moveChild(long newParentId, String newRole, int newIndex, long childId);

  ITree addNewChild(long parentId, String role, int index, long childId, IConcept concept);
  ITree addNewChildren(long parentId, String role, int index, long[] newIds, IConcept[] concepts);

  ITree deleteNode(long nodeId);
  ITree deleteNodes(long[] nodeIds);
}

File: model-client/src/main/kotlin/org/modelix/model/api/ITree.kt
package org.modelix.model.api

import java.util.stream.LongStream

interface ITree {
    fun visitChanges(oldVersion: ITree?, visitor: ITreeChangeVisitor?)
    fun containsNode(nodeId: Long): Boolean
    fun getProperty(nodeId: Long, role: String?): String?
    fun getChildren(parentId: Long, role: String?): LongStream?
    fun getConcept(nodeId: Long): IConcept?
    fun getParent(nodeId: Long): Long
    fun getRole(nodeId: Long): String?
    fun setProperty(nodeId: Long, role: String?, value: String?): ITree?
    fun getReferenceTarget(sourceId: Long, role: String?): INodeReference?
    fun setReferenceTarget(sourceId: Long, role: String?, target: INodeReference?): ITree?
    fun getReferenceRoles(sourceId: Long): Iterable<String?>?
    fun getPropertyRoles(sourceId: Long): Iterable<String?>?
    fun getChildRoles(sourceId: Long): Iterable<String?>?
    fun getAllChildren(parentId: Long): LongStream?
    fun moveChild(newParentId: Long, newRole: String?, newIndex: Int, childId: Long): ITree?
    fun addNewChild(parentId: Long, role: String?, index: Int, childId: Long, concept: IConcept?): ITree?
    fun addNewChildren(parentId: Long, role: String?, index: Int, newIds: LongArray?, concepts: Array<IConcept?>?): ITree?
    fun deleteNode(nodeId: Long): ITree?
    fun deleteNodes(nodeIds: LongArray?): ITree?

    companion object {
        const val ROOT_ID = 1L
        const val DETACHED_NODES_ROLE = "detached"
    }
}
